PAM: Add passkey preflight operation

justin-stephenson opened the pull request:
PIN attempts message is only shown when <= 3

```
justin@fedora:~$ su - pkuser
Insert your passkey device, then press ENTER.
Enter PIN:
You have 3 PIN attempts remaining

```

In addition, refactored a lot of PAM passkey child related code

justin-stephenson commented:
> I did not check the code, but went straight to testing. There is one thing that caught my attention. When the number of attempts remaining is less than 3 the user is supposed to get that number before asking for the PIN, so that they are aware that they are in a risky situation. Something like:
> 
> ```
> $ su - iker@ipa.test
> Insert your passkey device, then press ENTER.
> You have 3 PIN attempts remaining
> Enter PIN:
> su: Authentication failure
> ```
> 
> In general terms, passkey preflight should be run during preauth stage. I don't know how hard it would be to implement it this way but it would make more sense.

It is run during preauth stage, I just had to move around the pam_sss code which sends the message to the console. Please try again.

ikerexxe commented on src/sss_client/sss_cli.h:
```diff
@@ -552,11 +552,13 @@ enum response_type {
+                            * data including number of PIN attempts remaining.
                             * @param
                             *   - prompt_pin
+                            *   - attempts
```
Can you also add the attempts to `SSS_PAM_PASSKEY_KRB_INFO` so that I can call `decode_pam_passkey_msg()` and get this information alongside other data that is necessary for the GUI to show it?

ikerexxe replied:
This idea sounds good, but in that case, can we have all preflight data (including the prompt for the PIN) in the same PAM response type? It would make more sense and that would allow to check for this data there and avoid checking SSS_PAM_PASSKEY_INFO, SSS_PAM_PASSKEY_KRB_INFO and SSS_PAM_PASSKEY_PREFLIGHT when all information is available in just one place.

justin-stephenson replied:
>> Both of those structures work for me as they are available for the calling context of the JSON message. But can we have everything structured properly. I have the impression that we "throw" variables there without any context and this is difficult to understand and maintain. What I mean, is that we could embed all passkey data in a structure for passkey inside of struct pam_auth_req or struct pam_ctx. This way we know that this is for this authentication method and not for anything else.

Yes no problem. Okay I'll try this approach first.

> Will this happen always? Or just sometimes? What is the use case that you have in your mind?

It depends where in the pre-auth/auth flow your code executed to retrieve this preflight data. Do you have code you can point me to so I can check if this will be a problem? Currently this preflight operation is performed towards the end of pam_reply() during preauth (PAM -> pam_sss client), executing the passkey child is an async operation.

ikerexxe replied:
Probably this will need to happen sooner as the JSON message is also generate in pam_reply(). generate_json_auth_message() is actually called [here](https://github.com/ikerexxe/sssd/blob/gdm-krb5-auth-sel/src/responder/pam/pamsrv_cmd.c#L1556), and by that moment the information should be already available.

justin-stephenson replied:
Okay PR is updated now.

The preflight output is now stored in the following struct inside `pam_auth_req`.

```
struct pam_preflight_data {
    int attempts;
    bool pin_required;
    bool obtained;
};
```

pam_reply() gets called twice here in the preauth phase, you just need to check if the 'obtained' boolean is true then that means passkey child preflight operation has completed and the preflight data was retrieved so you can get the data.
```

        if (preq->pam_pf_data->obtained) {
            DEBUG(SSSDBG_TRACE_FUNC, "JS-attempts, [%d]\n", preq->pam_pf_data->attempts);
            DEBUG(SSSDBG_TRACE_FUNC, "JS-pin_required [%s]\n", preq->pam_pf_data->pin_required ? "True" : "False");
        }


/var/log/sssd/sssd_pam.log:1075:(2025-06-25 12:01:54): [pam] [pam_reply] (0x0400): [CID#6] JS-attempts, [6]
/var/log/sssd/sssd_pam.log:1076:(2025-06-25 12:01:54): [pam] [pam_reply] (0x0400): [CID#6] JS-pin_required [True]
```

ikerexxe reviewed (commented):
Ok, now the testing worked as expected:
```
su - iker@ipa.test
Insert your passkey device, then press ENTER.
You have 3 PIN attempts remaining

Enter PIN:
su: Authentication failure
```

I didn't check all the code, but there is something that caught my eye. See inline.

justin-stephenson commented:
@ikerexxe Please confirm these are the appropriate backport labels to use also

alexey-tikhonov commented:
> @ikerexxe Please confirm these are the appropriate backport labels to use also

What RHEL versions do you target?

ikerexxe commented on Makefile.am:
```diff
@@ -2603,7 +2608,8 @@ pam_srv_tests_LDADD = \
 if BUILD_PASSKEY
-     pam_srv_tests_SOURCES += src/responder/pam/pamsrv_passkey.c
+     pam_srv_tests_SOURCES += src/responder/pam/pamsrv_passkey.c \
+							  src/krb5_plugin/passkey/passkey_utils.c
```
Indentation issue

ikerexxe commented on src/responder/pam/pamsrv_passkey.h:
```diff
@@ -36,14 +36,19 @@ enum passkey_user_verification {
+    PAM_PASSKEY_OP_INVALID
+};
+
+struct pam_preflight_data {
```
Can we call it `pam_passkey_data`? The idea would be to include all passkey related data in the same structure

ikerexxe replied:
> Okay what other data should be stored in `pam_passkey_data` ? I still think it makes sense to separate the preflight data somehow to make it clear it is preflight specific data.

Nothing for the moment, but I see a pattern where this data will grow and we'll need to accommodate it.

In any case, this is just an opinion so if you want to keep this as it is I'm also fine with it. We can always refactor later on when the need is there.

ikerexxe commented on src/responder/pam/pamsrv_passkey.h:
```diff
@@ -36,14 +36,19 @@ enum passkey_user_verification {
+struct pam_preflight_data {
+    int attempts;
+    bool pin_required;
+    bool obtained;
```
Since we are changing the structure name I'd also change this to `preflight_obtained` or `preflight_executed`

ikerexxe reviewed (commented):
Just some minor comments. By the way, I forgot to mention that I tested it and it works as expected

alexey-tikhonov commented:
/gemini review

alexey-tikhonov commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -327,6 +338,482 @@ errno_t pam_passkey_get_mapping_recv(TALLOC_CTX *mem_ctx,
+                                    pk_check_ctx->pd,
+                                    pk_check_ctx->op,
+                                    timeout,
+                                    debug_libfido2,
+                                    verification,
+                                    data);
```
/gemini see comment above.

justin-stephenson commented:
> @justin-stephenson, could you please rebase (again)? There is a trivial conflict because #7991 was merged. Unfortunately, this means 'sssd-2-9' will require explicit backport PR.

Rebased.

alexey-tikhonov commented on contrib/sssd.spec.in:
```diff
@@ -960,6 +963,7 @@ install -D -p -m 0644 %{SOURCE1} %{buildroot}%{_sysusersdir}/sssd.conf
 %{_datadir}/sssd/krb5-snippets/sssd_enable_passkey
 %{_udevrulesdir}/90-sssd-token-access.rules
 %config(noreplace) %{_sysconfdir}/krb5.conf.d/sssd_enable_passkey
+%{_mandir}/man5/sssd-passkey.5*
```
Please add `:packaging:` RN to the commit message.

justin-stephenson replied:
Done.

justin-stephenson commented on src/man/sssd-passkey.5.xml:
```diff
@@ -0,0 +1,128 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Done.

alexey-tikhonov commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -890,3 +910,47 @@ verify_assert_data(struct passkey_data *data)
 }
+
+errno_t
+preflight(struct passkey_data *data, int timeout)
```
`select_authenticator()` /  `preflight()` could take pointer to `const struct passkey_data`

alexey-tikhonov commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -244,3 +244,25 @@ get_device_options(fido_dev_t *dev, struct passkey_data *_data)
 }
+
+errno_t
+get_device_pin_retries(fido_dev_t *dev, struct passkey_data *data,
```
`data` could be a pointer to `const struct passkey_data`

alexey-tikhonov commented on src/sss_client/pam_message.h:
```diff
@@ -66,6 +66,7 @@ struct pam_items {
     char *first_factor;
     char *passkey_key;
     char *passkey_prompt_pin;
+    uint32_t attempts;
```
This is pretty messy structure :-/

Would it make sense to add 'passkey_' or 'passkey_pin_' prefix to 'attempts'?
Or is this going to be reused outside of passkey context?

alexey-tikhonov commented on src/sss_client/pam_sss.c:
```diff
@@ -1338,17 +1338,20 @@ static int eval_response(pam_handle_t *pamh, size_t buflen, uint8_t *buf,
                 }
                 break;
             case SSS_PAM_PASSKEY_INFO:
+                memcpy(&pi->attempts, &buf[p], sizeof(uint32_t));
```
`safealign_memcpy()`

justin-stephenson replied:
Fixed.

alexey-tikhonov commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -327,6 +338,482 @@ errno_t pam_passkey_get_mapping_recv(TALLOC_CTX *mem_ctx,
+    enum passkey_child_op op;
+};
+
+int passkey_child_execute(TALLOC_CTX *mem_ctx,
```
What is the point of forward declaration (with a different return type, btw)?

justin-stephenson commented:
> I've left a bunch of (pretty cosmetic) comments inline.
> 
> But I'm only reading code, I can't actually test this.
> 
> Thus I'll add @madhuriupadhye as additional reviewer (to actually test this and also fix passkey tests that are broken by this PR).

Thank you for the review. I addressed comments related to my code, and will let Iker address his commits when he returns from PTO.

justin-stephenson commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -327,6 +338,482 @@ errno_t pam_passkey_get_mapping_recv(TALLOC_CTX *mem_ctx,
+        goto done;
+    }
+
+    tevent_req_set_callback(req, passkey_child_execute_done, pk_check_ctx);
```
Disregard my comment above. I see the issue now. I will fix it.

justin-stephenson replied:
I renamed `passkey_child_execute_done` to `passkey_child_get_user_data_done`

ikerexxe commented on src/passkey_child/passkey_child.h:
```diff
@@ -342,7 +346,7 @@ authenticate(struct passkey_data *data);
  */
 errno_t
-select_authenticator(struct passkey_data *data, fido_dev_t **_dev,
+select_authenticator(struct passkey_data *data, int timeout, fido_dev_t **_dev,
```
I want to be able to run unit tests for this function

ikerexxe commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -28,11 +28,11 @@
     errno_t ret;
 
-    for (int i = 0; i < TIMEOUT; i += FREQUENCY) {
+    for (int i = 0; i < timeout; i += FREQUENCY) {
```
This looks good. @justin-stephenson how do you want to handle this change? I see you already modified some of the code I provided, so can you also take care of this?

ikerexxe replied:
Done in a separate commit

alexey-tikhonov commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -262,6 +264,16 @@ parse_arguments(TALLOC_CTX *mem_ctx, int argc, const char *argv[],
             break;
+        case 'p':
+            if (data->action != ACTION_NONE) {
+                fprintf(stderr, "\nActions are mutually exclusive and should" \
```
ok

alexey-tikhonov commented:
> > Do I understand correctly that the only point of `passkey: timeout argument refactor` patch is to use 'timeout = 1' for `preflight()`?
> > I think it's not enough.
> > 'sssd_pam' should provide `timeout` as a command line argument to 'passkey_child' and this (perhaps decreased by 1) should be used instead of `TIMEOUT` define (see 'p11_child' as example).
> 
> Unless you plan to provide a way for this to be configurable 

It is already configurable via `sssd.conf::passkey_child_timeout` option.

> I don't see any difference between the actual implementation and your proposal.

The issue is that 'passkey_child' doesn't use it.

ikerexxe commented:
There are two distinct timeouts happening here: the timeout you're referring to controls how long the PAM responder will wait for the `passkey_child` process to complete. The other, separate timeout determines how much time the user has to physically connect their FIDO2 token.

alexey-tikhonov commented:
> There are two distinct timeouts happening here: the timeout you're referring to controls how long the PAM responder will wait for the `passkey_child` process to complete. The other, separate timeout determines how much time the user has to physically connect their FIDO2 token.

They should be related. At the very least 2nd can't be greater than 1st.
Take 'p11_child' as an example: all internal child timeouts should be less than governing sssd_pam timeout.

ikerexxe commented:
Agreed to tackle the timeout topic on a different PR.

ikerexxe commented:
CI, and more specifically the system tests, are in the red, and I believe this is due to our changes. I have asked Madhuri to review them.
